### PR TITLE
ENH/BUG Bootstrap clone rebased

### PR DESCRIPTION
**Pull request by josef-pkt**
this is a rebased version of #4676

### CODE REVIEW

**Comment by josef-pkt** [comment]
The notebook contains the output and needs to be removed, commit "DOC add example noexog GLLM notebook"

File: statsmodels/miscmodels/tests/test_generic_mle.py [hunk 1]
```diff
@@ -149,3 +149,79 @@ def setup_class(cls):
...
+        cls.pdf_a = pdf_a
+        cls.pdf_b = pdf_b
+
+    def test_fit(self):
+        llh_noexog = TwoPeakLLHNoExog(self.X,
+                                      signal=self.pdf_a,
+                                      background=self.pdf_b)
+
+        res = llh_noexog.fit()
+        assert np.allclose(res.params, self.params, rtol=1e-1)
```
Comment (josef-pkt) [inline statsmodels/miscmodels/tests/test_generic_mle.py]: use assert_allclose

**Comment by josef-pkt** [comment]
aside: why are df_resid, df_model nan?

notebook: I prefer not to have dedented closing parenthesis.

**Comment by josef-pkt** [comment]
AFAICS, notebook doesn't show usage of bootstrap cov_params/bse

**Comment by josef-pkt** [comment]
The handling of exog=None in bootstrap looks good. It's just the minimal change without getting into the refactoring parts, see #4679 and discussion in original PR

**Comment by josef-pkt** [comment]
we can squash into 3 parts: 
changes to code
notebook
unit test

**Comment by josef-pkt** [comment]
df_model, df_resid are nan:

there is an inconsistency between GenericLikelihoodModel and the results class

model.initialize has, i.e. attribute is always set
```
        if self.exog is not None:
            # assume constant
            er = np_matrix_rank(self.exog)
            self.df_model = float(er - 1)
            self.df_resid = float(self.exog.shape[0] - er)
        else:
            self.df_model = np.nan
            self.df_resid = np.nan
```

`GenericLikelihoodModelResults.__init__` uses hasattr, i.e. we never get to the else part to set df_model equal to len(params)

```
        if hasattr(model, 'df_model'):
            self.df_model = model.df_model
        else:
            self.df_model = len(mlefit.params)
            # retrofitting the model, used in t_test TODO: check design
            self.model.df_model = self.df_model

        if hasattr(model, 'df_resid'):
            self.df_resid = model.df_resid
        else:
            self.df_resid = self.endog.shape[0] - self.df_model
            # retrofitting the model, used in t_test TODO: check design
            self.model.df_resid = self.df_resid
```

df_resid does not affect params_table by default using use_t=False.
It shows up in summary where formatting is fixed in this PR
It is also required for t_test (but I don't remember if we only need the attribute or if it needs to have a non-nan value.)
It should also be needed for nested hypothesis test when df_constraints = diff of df_resid, but that might not be implemented for GenericLikelihoodModelResults.

File: statsmodels/iolib/summary.py [hunk 2]
```diff
@@ -307,16 +312,16 @@ def summary_top(results, title=None, gleft=None, gright=None, yname=None, xname=
...
           #TODO: check when we have non-integer df
-          ('Df Residuals:', lambda: ["%#6d" % results.df_resid]),
-          #('Df resid:', lambda: [results.df_resid]),
-          #('df resid:', lambda: [results.df_resid]), #check capitalization
-          ('Log-Likelihood:', lambda: ["%#8.5g" % results.llf]) #doesn't exist for RLM - exception
-          #('Method:', lambda: [???]), #no default for this
-          ])
+          ('Df Residuals:', lambda: [num_to_str(results.df_resid)]),
+          # ('Df resid:', lambda: [results.df_resid]),
+          # ('df resid:', lambda: [results.df_resid]), #check capitalization
```
Comment (jbrockmendel) [inline statsmodels/iolib/summary.py]: Commented out versions necessary?

File: statsmodels/miscmodels/tests/test_generic_mle.py [hunk 3]
```diff
@@ -149,3 +149,79 @@ def setup_class(cls):
...
+    endog_names = ['alpha']
+
+    def __init__(self, endog, exog=None, signal=None, background=None,
+                 *args, **kwargs):
+        # assume we know the shape + location of the two components,
+        # so we re-use their PDFs here
+        self.signal = signal
+        self.background = background
+        super(TwoPeakLLHNoExog, self).__init__(endog=endog, exog=exog,
+                                         *args, **kwargs)
```
Comment (jbrockmendel) [inline statsmodels/miscmodels/tests/test_generic_mle.py]: Indentation

**Comment by josef-pkt** [comment]
I want to remove the doc notebook from this PR. My guess is that there is something wrong in the model or likelihood. The parameter estimates are fine and bootstrap works, but the hessian doesn't look correct for this problem.
Reparameterization to us n_signal and n_total as parameters makes the two parameters less correlated, using fraction and n_total would make it them pretty orthogonal.
However, n_total is just a estimate of nobs with very small bse 3e-5 in bootstrap but much larger bse=32 according to hessian.

As a unit test case the example is fine, but I don't want to provide as a user example in the docs, when something is not clear or appropriate in the example.

**Comment by josef-pkt** [comment]
update: I added `nloglikeobs` method to get hc0, which is much closer to, even smaller than, bse_boot for n_total

```
res.bsejhj
array([  8.85731641e+00,   5.18898493e-07])

res_hc0.bse   # using fit(cov_type='hc0')
array([  8.85731641e+00,   5.17198070e-07])

bse_boot
array([  8.92083745e+00,   2.99404115e-05])
```

**Comment by josef-pkt** [comment]
checking current unit test

statsmodels.miscmodels.tests.test_generic_mle  TestMyPareto1 which uses constraints is None
MyPareto has no exog
if I call bootstrap, then I get a different error
`AttributeError: 'MyPareto' object has no attribute 'fixed_params'`

my guess is that this is related to missing `_init_keys` or the current bootstrap `cloneattr`
The constrained handling is part of the user class, i.e. MyPareto.

My first tries to fix MyPareto to get bootstrap with constraints to work are failing.
Also in bootstrap I get a lot of convergence failures or HessianInversionWarnings.
It looks like recreating the model in bootstrap is not picking up the constraint.

**Comment by josef-pkt** [comment]
bootstrap for scale parameter in MyPareto with loc restriction looks not very normal
e.g.  robust mad looks pretty different from bse_boots
asymptotic bse in model for scale is 4.89e-06

image

**Comment by josef-pkt** [comment]
I copied the notebook to https://gist.github.com/josef-pkt/892fd3a6048cf55790a66e152ba3cc7d
and will delete it from here 
It disappeared from this PR after interactive rebase

now squashed to 3 commits. 
- bug-fix
- allow nan in summary
- unit tests

**Comment by josef-pkt** [comment]
All green,

I'm leaving the writing of nice examples for another day. E.g. expanding on MyPareto

File: statsmodels/base/model.py [hunk 4]
```diff
@@ -2213,13 +2213,17 @@ def bootstrap(self, nrep=100, method='nm', disp=0, store=1):
         """
         results = []
         print(self.model.__class__)
-        hascloneattr = True if hasattr(self, 'cloneattr') else False
+        hascloneattr = True if hasattr(self.model, 'cloneattr') else False
```
Comment (jbrockmendel) [inline statsmodels/base/model.py]: Is cloneattr documented somewhere?

File: statsmodels/iolib/summary.py (hunk 2 above)
Comment (jbrockmendel) [reply statsmodels/iolib/summary.py]: Commented out versions necessary?

File: statsmodels/miscmodels/tests/test_generic_mle.py [hunk 5]
```diff
@@ -149,3 +150,76 @@ def setup_class(cls):
...
+
+    def loglike(self, params):        # pylint: disable=E0202
+        return -self.nloglike(params)
+
+    def nloglike(self, params):
+        endog = self.endog
+        return self.nlnlike(params, endog)
+
+    def nlnlike(self, params, endog):
```
Comment (jbrockmendel) [inline statsmodels/miscmodels/tests/test_generic_mle.py]: Why defining both this and nloglike?

Comment (josef-pkt) [reply statsmodels/miscmodels/tests/test_generic_mle.py]: small advantage, nlnlike can be used with different endog

but essentially I didn't rewrite the example in the unit test.
I have a partial rewrite in my notebook, for the future.

File: statsmodels/base/model.py (hunk 4 above)
Comment (josef-pkt) [reply statsmodels/base/model.py]: No, it's not advertised.
discussion in the original PR

File: statsmodels/iolib/summary.py (hunk 2 above)
Comment (josef-pkt) [reply statsmodels/iolib/summary.py]: I don't know. postponed and new open issue for reviewing GenericLikelihoodModel.
this PR makes the minimal changes to fix the bug to get it to work.

**Pull request merged**